packages/subgraph/src/utils/tokens/erc721.ts

```diff
@@ -138,97 +138,106 @@ export function handleERC721Action(
 
-  if (
-    functionSelector == ERC721_transferFrom ||
-    functionSelector == ERC721_safeTransferFromNoData
-  ) {
-    decodeABI = '(address,address,uint256)';
-  }
-
-  if (functionSelector == ERC721_safeTransferFromWithData) {
-    decodeABI = '(address,address,uint256,bytes)';
-    calldata = DECODE_OFFSET + calldata;
-  }
-
-  let decoded = ethereum.decode(decodeABI, Bytes.fromHexString(calldata));
-
-  if (!decoded) {
-    return;
-  }
+  if (!decodeABI) return false;
 
-  let tuple = decoded.toTuple();
-
-  let from = tuple[0].toAddress();
-  let to = tuple[1].toAddress();
-  let tokenId = tuple[2].toBigInt();
-
-  let daoId = dao.toHexString();
-
-  let transferId = getTransferId(
-    event.transaction.hash,
-    event.transactionLogIndex,
+  let calldata = getCalldata(functionSelector, data);
+  let decoded = ethereum.decode(
+    decodeABI as string,
+    Bytes.fromHexString(calldata)
+  );
+  if (!decoded) return false;
+
+  let transfer = createERC721Transfer(
+    decoded.toTuple(),
+    contract,
+    dao,
+    event,
+    proposalId,
     actionIndex
   );
 
-  let transfer = new ERC721Transfer(transferId);
-  transfer.from = from;
-  transfer.to = to;
-  transfer.dao = daoId;
-  transfer.token = contract.id;
-  transfer.tokenId = tokenId;
-  transfer.proposal = proposalId;
-  transfer.txHash = event.transaction.hash;
-  transfer.createdAt = event.block.timestamp;
-
-  if (from == dao && to == dao) {
+  if (transfer.from == dao && transfer.to == dao) {
     transfer.type = 'Withdraw';
-    transfer.save();
-    return;
-  }
-
-  // If from/to both aren't equal to dao, it means
-  // dao must have been approved for the `tokenId`
-  // and played the role of transfering between 2 parties.
-  if (from != dao && to != dao) {
+  } else if (transfer.from != dao && transfer.to != dao) {
+    // If from/to both aren't equal to dao, it means
+    // dao must have been approved for the `tokenId`
+    // and played the role of transfering between 2 parties.
     transfer.type = 'ExternalTransfer';
-    transfer.save();
-    return;
-  }
-
-  if (from != dao && to == dao) {
+  } else if (transfer.from != dao && transfer.to == dao) {
     // 1. some party `y` approved `x` tokenId to the dao.
     // 2. dao calls transferFrom as an action to transfer it from `y` to itself.
     transfer.type = 'Deposit';
-
     updateERC721Balance(
-      daoId,
-      token.toHexString(),
-      tokenId,
+      transfer.dao,
+      transfer.token,
+      transfer.tokenId,
       event.block.timestamp,
       TransferType.Deposit
     );
   } else {
     transfer.type = 'Withdraw';
-
     updateERC721Balance(
-      daoId,
-      token.toHexString(),
-      tokenId,
+      transfer.dao,
+      transfer.token,
+      transfer.tokenId,
       event.block.timestamp,
       TransferType.Withdraw
     );
   }
 
   transfer.save();
+  return true;
+}
+
+function determineDecodeABI(functionSelector: string): string | null {
```

Comment:
maybe rename to `determineERC721DecodeABI` since we have `determineERC1155DecodeABI` for the 1155?

Reply:
addressed in [16b8444](https://github.com/aragon/osx/pull/472/commits/16b84446211c26f20054618fedbec9ad87504365)